File: tiny-projects/process_scheduler/scheduler.hs
import Data.Set

data Proc = Proc { procId :: Int
                 , startTime :: Int
                 , procLen :: Int
                 , waitTime :: Int
                 , readyTimes :: Int
                 , restWorks :: Int
                 } deriving (Show)
type PID = Int

seq2proc::(PID,Int,Int) -> Proc
seq2proc (pid,len,start) = Proc {procId=pid, startTime=start, procLen=len, waitTime=0, readyTimes=1, restWorks=len}

-- scheduler implimentation
data AbsProc = SingleProc Proc | CombinedProc [(Proc, Int)] deriving (Show)




type PIDState = (PID, Int) -- Int be the rest works
type PeriodSignal = ()
type Scheduler = (PeriodSignal -> PID)
type MetaScheduler = [Proc] -> Scheduler
type ExeSeq = [PID]

execute :: (Set PIDState) -> Scheduler -> ExeSeq
execute pstateSet scheduler procExeSeq = 
    if null $ filter (\(_,restworks) -> restworks > 0) pstateSet then
        reverse procExeSeq
    else
        let nextProc = (scheduler ()) 
            p = (head $ filter (\(nextproc,_)->True) pstateSet)
            newSet = Set.insert (fst p, (snd p) -1) (Set.delete p pstateSet)
        in
            execute newSet scheduler (nextProc : ProcExeSeq)
        
        
runmachine:: [Proc] -> MetaScheduler -> ExeSeq
runmachine procs metaScheduler =
    let getPidState = (\Proc {procId=pid, procLen=len} -> (pid,len))
    in
        execute (Set.fromList $ map getPidState procs) (metaScheduler procs) []

---------------
evalAWT :: [Proc] -> ExeSeq -> Double

File: tiny-projects/process_scheduler/scheduler.pl
/* 
proc( id,
      startTime,
      burstTime,
      waitTime ,
      readyTimes)
*/

makeScheduler(ProcList):-
nextToRun(A):-



runOnce(PID,Rest):-
    runningproc(PID,Burst),
    retract(runningproc(PID,Burst)),
    Rest is Burst - 1,
    assertz(runningproc(PID,Rest)).
increaseWaitTime(PID):-
    proc(PID,A,B,Wait,C),
    retract(proc(PID,_,_,_,_)),
    NWait is Wait+1,
    assertz(PID,A,B,NWait,C).
increaseReadyTimes(PID):-
    proc(PID,A,B,C,Ready),
    retract(proc(PID,_,_,_,_)),
    NR is Ready+1,
    assertz(PID,A,B,C,NR).


execute(ProcSeq,ProcDone):-
    nextToRun(P,ProcDone),
    (P == 0 ->
        ProcSeq = [];
        runOnce(P,Rest),
        forall(proc(WP,_,_,_,_),WP /= P,increaseWaitTime(WP)),
        (ProcSeq = [Last|_] ->
            increaseReadyTimes(Last);true),
        (Rest>0 ->
            execute([P|ProcSeq],fail);
            execute([P|ProcSeq],true)))

initialProcs([]).
initialProcs([proc(ID,_,Burst,_,_)|Ps]):-
    assertz(runningproc(ID,Burst)),
    initialProcs(Ps).

runMachine(ProcList,ProcSeq):-
    makeScheduler(ProcList),
    initialProcs(ProcList),!,
    execute(Result),
    reverse(Result,ProcSeq).

sumWaitTime([],0).
sumWaitTime([P|Ps],R):-
    proc(P,_,_,Wait,Ready),
    sumWaitTime(Ps,NR),
    R is Wait/Ready + NR.

evalAWT(ProcList,Result):-
    sumWaitTime(ProcList,TotalWaitTime),
    length(ProcList,Len),
    Result is TotalWaitTime/Len.
